velox/common/encode/Base64.cpp

```diff
@@ -310,81 +286,66 @@ void Base64::decode(const char* data, size_t size, char* output) {
   Base64::decode(data, size, output, out_len);
 }
 
-uint8_t Base64::Base64ReverseLookup(
-    char p,
-    const Base64::ReverseIndex& reverse_lookup) {
-  auto curr = reverse_lookup[(uint8_t)p];
-  if (curr >= 0x40) {
-    throw Base64Exception(
-        "Base64::decode() - invalid input string: invalid characters");
-  }
-
-  return curr;
-}
-
 size_t
 Base64::decode(const char* src, size_t src_len, char* dst, size_t dst_len) {
-  return decodeImpl(src, src_len, dst, dst_len, kBase64ReverseIndexTable, true);
+  return decodeImpl(src, src_len, dst, dst_len, kBase64ReverseIndexTable);
 }
 
-// static
-size_t
-Base64::calculateDecodedSize(const char* data, size_t& size, bool withPadding) {
+size_t Base64::calculateDecodedSize(const char* data, size_t& size) {
   if (size == 0) {
     return 0;
   }
 
-  auto needed = (size / 4) * 3;
-  if (withPadding) {
-    // If the pad characters are included then the source string must be a
-    // multiple of 4 and we can query the end of the string to see how much
-    // padding exists.
-    if (size % 4 != 0) {
-      throw Base64Exception(
+  // Check if the input data is padded
+  if (isPadded(data, size)) {
+    /// If padded, ensure that the string length is a multiple of the encoded
+    /// block size
+    if (size % kEncodedBlockSize != 0) {
+      throw EncoderException(
           "Base64::decode() - invalid input string: "
-          "string length is not multiple of 4.");
+          "string length is not a multiple of the encoded block size.");
     }
 
-    auto padding = countPadding(data, size);
+    auto needed = (size * kBinaryBlockSize) / kEncodedBlockSize;
+    auto padding = EncoderBase::countPadding(data, size);
     size -= padding;
-    return needed - padding;
-  }
 
-  // If padding doesn't exist we need to calculate it from the size - if the
-  // size % 4 is 0 then we have an even multiple 3 byte chunks in the result
-  // if it is 2 then we need 1 more byte in the output.  If it is 3 then we
-  // need 2 more bytes in the output.  It should never be 1.
-  auto extra = size % 4;
-  if (extra) {
-    if (extra == 1) {
-      throw Base64Exception(
-          "Base64::decode() - invalid input string: "
-          "string length cannot be 1 more than a multiple of 4.");
+    // Adjust the needed size for padding
+    return needed -
+        ceil((padding * kBinaryBlockSize) /
+             static_cast<double>(kEncodedBlockSize));
+  } else {
+    // If not padded, Calculate extra bytes, if any
+    auto extra = size % kEncodedBlockSize;
+    auto needed = (size / kEncodedBlockSize) * kBinaryBlockSize;
+
+    // Adjust the needed size for extra bytes, if present
+    if (extra) {
+      if (extra == 1) {
+        throw EncoderException(
+            "Base64::decode() - invalid input string: "
+            "string length cannot be 1 more than a multiple of 4.");
+      }
+      needed += (extra * kBinaryBlockSize) / kEncodedBlockSize;
```

Comment:
I don't follow why this change is needed. Maybe a separate PR could be opened if changes to Base64 encoding are needed?

Reply:
@pramodsatya Base64 had logical issues decoding the characters if they were not padded.

While trying to make common encoder logic, it also had to be fixed, that's the reason for the above change.

You can run the additional test cases below to reproduce the bug.

```
  // Check encoded strings without padding
  EXPECT_EQ("a", fromBase64("YQ"));
  EXPECT_EQ("ab", fromBase64("YWI"));
  EXPECT_EQ("abcd", fromBase64("YWJjZA"));
```

I didn't create a separate PR because the changes are dependent, and I don't want to open multiple dependent PRs.
Please share your thoughts.

Reply:
Thanks for the clarification @Joe-Abraham, LGTM.